### PR TITLE
[stable/postgresql] Support for BoM and easy image re-location

### PR DESCRIPTION
**Pull request by jhvhs**
#### What this PR does / why we need it:
Add support for the [image referencing convention](https://github.com/helm/helm/issues/7154), the main goal of which is to enable the end user to easily get a list of images used in the chart, and to be able to simplify the use of the chart with re-located images, both on its own, as well as a sub-chart.
Please see https://github.com/helm/helm/issues/7154 for more details.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

File: stable/postgresql/templates/statefulset-slaves.yaml [hunk 1]
```diff
@@ -70,8 +70,7 @@ spec:
       initContainers:
       {{- if or (and .Values.volumePermissions.enabled .Values.persistence.enabled) .Values.shmVolume.enabled }}
         - name: init-chmod-data
-          image: {{ template "postgresql.volumePermissions.image" . }}
-          imagePullPolicy: "{{ .Values.volumePermissions.image.pullPolicy }}"
+{{ include "postgresql.registryImage" (dict "image" .Values.images.volumePermissions "values" .Values) | indent 10 }}
```
Comment (javsalgar) [inline stable/postgresql/templates/statefulset-slaves.yaml]: Could you please use `{{-` and `nindent` for consistency and allowing  tabbing (it is the standard we use in the rest of indents)?.

**Review by javsalgar — COMMENTED** [review]
Hi,
Thanks for the Pull Request. After checking it, allow me to raise some concerns that we have with the proposed changes.

* Looking at https://github.com/helm/helm/issues/7154 we do not see any reference to a design document or a discussion on the current design. As this is a major change, we would like to know how definitive this convention would be. Is it still open for discussion? If we are still testing this convention, could we use a different chart that drives less usage?

* In our case, the use of global.imageRegistry was introduced for allowing a better use experience in clouds like AWS and Azure, where they use their own image registries (different from docker.io). Instead of having to define the registry in each of the images (including sub-charts), the use of global.imageRegistry would fix it by simply putting --set global.imageRegistry example.com . In this PR, images.postgresql.registry would have more priority than global.imageRegistry, resulting in a poor user experience as they would have to set global.imageRegistry and unset images.postgresql.registry . Because of this, I believe that images.postgresql.registry  should be null by default in values.yaml

**Comment by jhvhs** [comment]
> * In our case, the use of global.imageRegistry was introduced for allowing a better use experience in clouds like AWS and Azure, where they use their own image registries (different from docker.io). Instead of having to define the registry in each of the images (including sub-charts), the use of global.imageRegistry would fix it by simply putting --set global.imageRegistry example.com . In this PR, images.postgresql.registry would have more priority than global.imageRegistry, resulting in a poor user experience as they would have to set global.imageRegistry and unset images.postgresql.registry . Because of this, I believe that images.postgresql.registry  should be null by default in values.yaml

We added settings for `image.*.registry` and `imageRegistry` to allow more flexibility, especially when overriding values in sub-charts, so that the registry and/or the namespace could be overridden either for a single image, or for all images in the chart, but without affecting other charts.

**Comment by jhvhs** [comment]
> * Looking at [helm/helm#7154](https://github.com/helm/helm/issues/7154) we do not see any reference to a design document or a discussion on the current design. As this is a major change, we would like to know how definitive this convention would be. Is it still open for discussion? If we are still testing this convention, could we use a different chart that drives less usage?

A design document is being produced right now, and we have been (and still are) actively soliciting feedback on this proposal based off its current description in the issue. I don't expect the interface to change much.

Nevertheless, we have added backwards compatibility with the original values.yaml.

**Comment by jhvhs** [comment]
The design document is available [here](https://docs.google.com/document/d/1zsuq_iOWCrwA2tnB0Ka0E4ewyr5cZIoHXLD6W24O2gE/edit?usp=sharing)

**Comment by javsalgar** [comment]
I've taken a look and in principle the changes look good to me. However, this will likely break all of the testing automations that we have in place so we would require some time to adapt our pipeline. I will update this ticket when we are ready for it.

**Comment by jhvhs** [comment]
@javsalgar The changes should be backwards compatible, and we have reflected that in the version bump.

**Comment by jhvhs** [comment]
@javsalgar We've added a test suite to demonstrate and verify the template helper behaviour

**Comment by jhvhs** [comment]
@javsalgar Let us know if you think the template helper tests should stay included in this repo.
They really helped us to develop the template helpers.

**Comment by davidkarlsen** [comment]
Chart is deprecated: https://github.com/helm/charts/pull/21408 - please use new repo for issues

**Pull request closed without merging**